File: 287-find-the-duplicate-number/find-the-duplicate-number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int a = 0;
        int b = 0;
        while (true) {
            a = nums[a];
            b = nums[nums[b]];
            if (a == b) {
                break;
            }
        }
        a = 0;
        while (nums[a] != nums[b]) {
            a = nums[a];
            b = nums[b];
        }
        return nums[a];
    }
};

File: 287-find-the-duplicate-number/find-the-duplicate-number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        n = len(nums)
        l, r = 1, n - 1
        def check(val):
            count = 0
            for n in nums:
                if n <= val:
                    count += 1
            return count
        while l < r:
            mid = (l + r) // 2
            if check(mid) > mid:
                r = mid
            else:
                l = mid + 1
        return l